File: rest-entities/src/main/java/org/chronopolis/rest/lombok/Node.java
package org.chronopolis.rest.lombok;

import lombok.Data;
import lombok.EqualsAndHashCode;
import org.hibernate.annotations.NaturalId;

import javax.persistence.Entity;
import javax.persistence.OneToMany;
import java.util.Set;

/**
 * @author shake
 */
@Data
@Entity
public class Node extends PersistableEntity {

    @NaturalId
    @EqualsAndHashCode.Include
    private String username;

    private String password;
    private Boolean enabled;

    @OneToMany(mappedBy = "node")
    private Set<Replication> replications;

}

File: rest-entities/src/main/kotlin/org/chronopolis/rest/kot/entities/Node.kt
package org.chronopolis.rest.kot.entities

import org.chronopolis.rest.kot.entities.depositor.DepositorNode
import javax.persistence.CascadeType
import javax.persistence.Entity
import javax.persistence.OneToMany

/**
 * Representation of a Node in Chronopolis
 *
 * todo: lateinit Sets?
 * todo: remove password/enabled... maybe rename username
 *
 * @author shake
 */
@Entity
class Node(
        @OneToMany(mappedBy = "node")
        var replications: Set<Replication> = emptySet(),

        // @OneToMany(mappedBy = "node")
        // var restorations: Set<Restoration> = emptySet(),

        @OneToMany(mappedBy = "node", cascade = [CascadeType.ALL], orphanRemoval = true)
        var depositorDistributions: MutableSet<DepositorNode> = mutableSetOf(),

        var username: String = "",
        var password: String = "",
        var enabled: Boolean = true
) : PersistableEntity()